File: 869_reordered_power_of2/reordered_power_of2.cpp
/*
 * @Date: 2021-10-28 01:59:30
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-10-28 02:09:50
 */

#include <cassert>
#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

string countDigits(int n) {
  string cnt(10, 0);
  while (n) {
    ++cnt[n % 10];
    n /= 10;
  }
  return cnt;
}

unordered_set<string> powerOf2Digits = []() {
  unordered_set<string> res{};
  for (int n = 1; n <= 1e9; n <<= 1) {
    res.insert(countDigits(n));
  }
  return res;
}();

class Solution {
 public:
  bool reorderedPowerOf2(int n) { return powerOf2Digits.count(countDigits(n)); }
};

int main() {
  assert(Solution().reorderedPowerOf2(1) == true);
  assert(Solution().reorderedPowerOf2(10) == false);
  assert(Solution().reorderedPowerOf2(15) == false);
  assert(Solution().reorderedPowerOf2(16) == true);
  assert(Solution().reorderedPowerOf2(24) == false);
  assert(Solution().reorderedPowerOf2(46) == true);
  assert(Solution().reorderedPowerOf2(8208) == false);
  assert(Solution().reorderedPowerOf2(9474) == false);
  assert(Solution().reorderedPowerOf2(2147483647) == false);
}

File: golang/869_reordered_power_of2/reordered_power_of2.go
// Package main ...
package main

func countDigits(n int) (cnt [10]int) {
	for n > 0 {
		cnt[n%10]++
		n /= 10
	}
	return
}

var powerOf2Digits = map[[10]int]bool{}

func init() {
	for n := 1; n <= 1e9; n <<= 1 {
		powerOf2Digits[countDigits(n)] = true
	}
}

func reorderedPowerOf2(n int) bool {
	return powerOf2Digits[countDigits(n)]
}

func main() {
	assert := func(result bool) {
		if !result {
			panic("assert failed")
		}
	}
	assert(reorderedPowerOf2(1) == true)
	assert(reorderedPowerOf2(10) == false)
	assert(reorderedPowerOf2(16) == true)
	assert(reorderedPowerOf2(24) == false)
	assert(reorderedPowerOf2(46) == true)
	assert(reorderedPowerOf2(8208) == false)
	assert(reorderedPowerOf2(9474) == false)
	assert(reorderedPowerOf2(15) == false)
	assert(reorderedPowerOf2(32) == true)
	assert(reorderedPowerOf2(8) == true)
	assert(reorderedPowerOf2(2147483647) == false)
	assert(reorderedPowerOf2(2147483648) == false)
}
